include/rmm/mr/device/detail/arena.hpp

```diff
@@ -32,87 +33,116 @@
 #include <memory>
 #include <mutex>
 #include <numeric>
+#include <optional>
 #include <set>
-#include <unordered_map>
 
 namespace rmm::mr::detail::arena {
 
-/// Minimum size of a superblock (256 KiB).
-constexpr std::size_t minimum_superblock_size = 1U << 18U;
-
 /**
- * @brief Represents a chunk of memory that can be allocated and deallocated.
- *
- * A block bigger than a certain size is called a "superblock".
+ * @brief Represents a contiguous region of memory.
  */
-class block {
+class memory_span {
  public:
   /**
-   * @brief Construct a default block.
-   */
-  block() = default;
-
-  /**
-   * @brief Construct a block given a pointer and size.
-   *
-   * @param pointer The address for the beginning of the block.
-   * @param size The size of the block.
+   * @brief Construct a default span.
    */
-  block(char* pointer, std::size_t size) : pointer_(pointer), size_(size) {}
+  memory_span() = default;
 
   /**
-   * @brief Construct a block given a void pointer and size.
+   * @brief Construct a span given a pointer and size.
    *
-   * @param pointer The address for the beginning of the block.
-   * @param size The size of the block.
+   * @param pointer The address for the beginning of the span.
+   * @param size The size of the span.
    */
-  block(void* pointer, std::size_t size) : pointer_(static_cast<char*>(pointer)), size_(size) {}
+  memory_span(void* pointer, std::size_t size) : pointer_{static_cast<char*>(pointer)}, size_{size}
+  {
+    RMM_LOGGING_ASSERT(pointer != nullptr);
+    RMM_LOGGING_ASSERT(size > 0);
+  }
 
   /// Returns the underlying pointer.
-  [[nodiscard]] void* pointer() const { return pointer_; }
+  [[nodiscard]] char* pointer() const { return pointer_; }
 
-  /// Returns the size of the block.
+  /// Returns the size of the span.
   [[nodiscard]] std::size_t size() const { return size_; }
 
-  /// Returns true if this block is valid (non-null), false otherwise.
-  [[nodiscard]] bool is_valid() const { return pointer_ != nullptr; }
+  /// Returns the end of the span.
+  [[nodiscard]] char* end() const
+  {
+    return pointer_ + size_;  // NOLINT(cppcoreguidelines-pro-bounds-pointer-arithmetic)
+  }
+
+  /// Returns true if this span is valid (non-null), false otherwise.
+  [[nodiscard]] bool is_valid() const { return pointer_ != nullptr && size_ > 0; }
```

Comment:
Consider moving `is_valid()` to the `block` class since that's where it is used.